File: 0983-minimum-cost-for-tickets/0983-minimum-cost-for-tickets.py
class Solution:
    
    def helper(self,days,costs,i,deepee):
        if i==len(days):
            return 0
        
        if deepee[i]!=-1: return deepee[i]
        ans1=costs[0] + self.helper(days, costs, i+1, deepee) 
        
        curr=i
        for index in range(curr,len(days)):
            if days[index]<days[i]+7:
                curr+=1
            else:
                break

        ans2=costs[1]+self.helper(days, costs, curr, deepee)
        curr=i
        for index in range(curr,len(days)):
            if days[index]<days[i]+30:
                curr+=1
            else:
                break

        ans3=costs[2]+self.helper(days, costs, curr, deepee)
        
        deepee[i] = min(ans1, ans2, ans3)
        
        return deepee[i]
        
    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        deepee=[-1 for i in range(len(days)+1)]
        return self.helper(days,costs,0,deepee)

File: 0983-minimum-cost-for-tickets/0983-minimum-cost-for-tickets.cpp
class Solution {
public:
    
    int minCost(int indx, vector<int> &days, vector<int> &costs, vector<int> &dp){
        
        int n = days.size();
        int i;
        
        //BASE CASE:
        if(indx >= n) return 0;
        
        if(dp[indx]!=-1) return dp[indx];
        
        //1-day pass
        int dayPass = costs[0] + minCost(indx+1, days, costs, dp);
        
        //7-day pass: skip 7 days if a 7-day pass is bought
        for(i = indx; i < n && days[i]<days[indx] + 7; i++);
        int weekPass = costs[1] + minCost(i, days, costs, dp);
        
        
        //30-Day pass
        for(i = indx; i<n && days[i]<days[indx]+30; i++);
        int monthPass = costs[2] + minCost(i, days, costs, dp);
        
        dp[indx] = min({dayPass, weekPass, monthPass});
        return dp[indx];
    }
    
    int mincostTickets(vector<int>& days, vector<int>& costs) {
        
        int n = days.size();
        vector<int> dp(n,-1);
        
        return minCost(0, days, costs, dp);
    }
};